snuba/clusters/cluster.py

```diff
@@ -0,0 +1,86 @@
+from typing import MutableMapping, Set
+
+from snuba import settings
+from snuba.clickhouse.native import ClickhousePool, NativeDriverReader
+from snuba.clickhouse.query import ClickhouseQuery
+from snuba.clusters.storage_sets import StorageSetKey, STORAGE_SETS
+from snuba.datasets.storages import StorageKey
+from snuba.reader import Reader
+
+
+class Cluster:
+    """
+    A cluster is responsible for managing a collection of database nodes.
+
+    Clusters are configurable, and will be instantiated based on user defined settings.
+
+    Each storage must be mapped to a cluster via a storage set, which defines
+    the storages that must be located on the same cluster.
+
+    A cluster provides a reader and Clickhouse connections that are shared by all
+    storages located on the cluster.
+
+    In future, clusters will also be responsible for co-ordinating commands that
+    need to be run on multiple hosts that are colocated within the same cluster.
+    The cluster will expose methods for:
+        - bootstrap
+        - migrate
+        - cleanup
+        - optimize
+    """
+
+    def __init__(self, host: str, port: int, http_port: int, storage_sets: Set[str]):
+        self.__storage_sets = storage_sets
+        self.__clickhouse_rw = ClickhousePool(host, port)
+        self.__clickhouse_ro = ClickhousePool(
+            host, port, client_settings={"readonly": True},
+        )
+        self.__reader: Reader[ClickhouseQuery] = NativeDriverReader(
+            self.__clickhouse_ro
+        )
+
+    def get_storage_sets(self) -> Set[StorageSetKey]:
+        return {StorageSetKey(storage_set) for storage_set in self.__storage_sets}
+
+    def get_clickhouse_rw(self) -> ClickhousePool:
+        return self.__clickhouse_rw
+
+    def get_clickhouse_ro(self) -> ClickhousePool:
+        return self.__clickhouse_ro
+
+    def get_reader(self) -> Reader[ClickhouseQuery]:
+        return self.__reader
```

Comment:
This comment is still valid:
https://github.com/getsentry/snuba/pull/860#discussion_r402033694